common4j/src/main/com/microsoft/identity/common/java/util/UrlUtil.java

```diff
@@ -56,16 +66,146 @@ public static String appendPathToURL(@NonNull final URL urlToAppend,
         final URIBuilder builder = new URIBuilder(urlToAppend.toString());
         final List<String> pathSegments = builder.getPathSegments();
 
-        final List<String> combinedPathSegments = new ArrayList<>();
-        combinedPathSegments.addAll(pathSegments);
+        final List<String> combinedPathSegments = new ArrayList<>(pathSegments);
 
-        for(final String path : pathSegmentsToAppend){
-            if (!StringUtil.isNullOrEmpty(path)){
+        for (final String path : pathSegmentsToAppend) {
+            if (!StringUtil.isNullOrEmpty(path)) {
                 combinedPathSegments.add(path);
             }
         }
 
         builder.setPathSegments(combinedPathSegments);
-        return builder.build().toString();
+        return builder.build().toURL();
     }
+
+
+    /**
+     * create url from given endpoint. return null if format is not right.
+     *
+     * @param endpoint url as a string
+     * @return URL object for this string
+     */
+    public static URL getUrl(@NonNull final String endpoint) {
+        final String methodName = ":getUrl";
+        try {
+            return new URL(endpoint);
+        } catch (final MalformedURLException e) {
+            Logger.errorPII(TAG + methodName, "URL is malformed", e);
+            return null;
+        }
+    }
+
+    /**
+     * Get URL parameters from final url.
+     *
+     * @param finalUrl String
+     * @return a map of url parameters.
+     */
+    @NonNull
+    public static Map<String, String> getUrlParameters(@Nullable final String finalUrl) {
+        if (StringUtil.isNullOrEmpty(finalUrl)){
+            return Collections.emptyMap();
+        }
+
+        final String methodName = ":getUrlParameters";
+        try {
+            final URI response = new URIBuilder(finalUrl).build();
+            final String fragment = response.getFragment();
+
+            if (!StringUtil.isNullOrEmpty(fragment) &&
+                    !urlFormDecode(fragment).isEmpty()) {
+                Logger.warn(TAG, "Received url contains unexpected fragment parameters.");
+                Logger.warnPII(TAG, "Unexpected fragment: " + response.getFragment());
+            }
+
+            if (StringUtil.isNullOrEmpty(response.getQuery())){
+                Logger.info(
+                        TAG + methodName,
+                        "URL does not contain query parameter"
+                );
+                return Collections.emptyMap();
+            }
+
+            return urlFormDecode(response.getQuery());
+        } catch (final URISyntaxException e) {
+            Logger.warn(TAG + methodName, "Failed to extract URL parameters");
+            Logger.errorPII(
+                    TAG + methodName,
+                    "Url is invalid",
+                    e
+            );
+            return Collections.emptyMap();
```

Comment:
Do we just want to throw the exception from here? I guess the question is that whether the entire token request should fail when this happens? What is the caller supposed to be with the empty map? Will they be able to proceed or are they supposed to throw their own exception if/once they realize the got back an empty map?

Reply:
Change the method such that it takes in URI only. this exception will be thrown somewhere else.